src/config/backends/odmdata.ts

```diff
@@ -0,0 +1,242 @@
+import * as _ from 'lodash';
+import { promises as fs } from 'fs';
+
+import { ConfigOptions, ConfigBackend } from './backend';
+import * as constants from '../../lib/constants';
+import log from '../../lib/supervisor-console';
+import { ODMDataError } from '../../lib/errors';
+
+/**
+ * A backend to handle ODMDATA configuration
+ *
+ * Supports:
+ * 	- {BALENA|RESIN}_HOST_CONFIG_odmdata_configuration = value | "value"
+ */
+
+export class Odmdata extends ConfigBackend {
+	private static bootConfigVarPrefix = `${constants.hostConfigVarPrefix}ODMDATA_`;
+	private static bootConfigPath = `/dev/mmcblk0boot0`;
+	private static bootConfigLockPath = `/sys/block/mmcblk0boot0/force_ro`;
+	private static supportedConfigs = ['configuration'];
+	private BYTE_OFFSETS = [1659, 5243, 18043];
+	private CONFIG_BYTES = [
+		0x0 /* Config Option #1 */,
+		0x1 /* Config Option #2 */,
+		0x6 /* Config Option #3 */,
+		0x7 /* Config Option #4 */,
+		0x2 /* Config Option #5 */,
+		0x3 /* Config Option #6 */,
+	];
+	private CONFIG_BUFFER = Buffer.from(this.CONFIG_BYTES);
+
+	public static bootConfigVarRegex = new RegExp(
+		'(?:' + _.escapeRegExp(Odmdata.bootConfigVarPrefix) + ')(.+)',
+	);
+
+	public async matches(deviceType: string): Promise<boolean> {
+		return deviceType.endsWith('-tx2');
+	}
+
+	public async getBootConfig(): Promise<ConfigOptions> {
+		// Get config buffer from bootConfigPath
+		const confBuffer = await this.readBootConfigPath();
+		// Parse ConfigOptions from bootConfigPath buffer
+		return this.parseOptions(confBuffer);
+	}
+
+	public async setBootConfig(opts: ConfigOptions): Promise<void> {
+		log.info('Attempting to configure ODMDATA.');
+		// Filter out unsupported options
+		const supportedOptions = _.pickBy(opts, (value, key) => {
+			if (!this.isSupportedConfig(key)) {
+				log.warn(`Not setting unsupported value: { ${key}: ${value} }`);
+				return false;
+			}
+			return true;
+		});
+		// Check that there is a configuration mode
+		if (!supportedOptions.configuration) {
+			log.info('No changes made to ODMDATA configuration.');
+			return;
+		}
+		// Check that confiugration mode is supported
+		const mode = parseInt(supportedOptions.configuration as string, 10);
+		if (mode < 1 || mode > 6) {
+			log.error(`Configuration mode of: ${mode} is not supported.`);
+			log.info('No changes made to ODMDATA configuration.');
+			return;
+		}
+		// Get the byte for this configuration mode
+		const byte = this.CONFIG_BYTES[mode - 1];
+		log.info(`Setting ODMDATA Configuration Mode #${mode}.`);
+		// Begin writing to bootConfigPath by unlocking RO access
```

Comment:
These comments aren't really very helpful. What I would like to see is a description of *why* we're doing what we're doing, not just a description of the code. The naming of functions et al should be enough to discern the what, but for example, I have no idea as a casual observer why we might be unlocking RO access. Unlocking access to what? Why do we need to unlock? etc